### PR TITLE
Generalize handling of arguments with binary value types

### LINKED ISSUE #452: reactions.get Facet fails to send request when requesting "full: true" and "file: id"
### Description

When trying to invoke ReactionFacet `reactions.get()` method passing an object with `file: id, full: true` properties, the request fails in a strange way. It appears to create a malformed request and actually fails prior to sending the request outside of the SDK code (apparently?). When removing the `full: true` property the request is made and responds as expect with either the discovered file, or a `{ "ok": false, "error": "file_not_found" }`. Adding back the `full: true` property to the object results in the request never taking place for some reason.

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [x] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Reproducible in:

`@slack/client` version: 3.15.0

node version: 8.9.0

OS version(s): macOS 10.13.2

#### Steps to reproduce:

1. Create simple app that imports `WebClient` from `@slack/client` module
2. New up `WebClient` with a `TOKEN`
3. Fire `web.reactions.get({ file: "any_string", full: true })` and view stack trace error

#### Expected result:

Return an instance of file object for a found match, or return 

```json
{
  "ok": false,
  "error": "file_not_found"
}
```

from Slack API.

#### Actual result:

Instead, the request fails before being sent, with the following stack trace:

```bash
_http_outgoing.js:640
    throw new TypeError('First argument must be a string or Buffer');
    ^
TypeError: First argument must be a string or Buffer
    at write_ (_http_outgoing.js:640:11)
    at ClientRequest.write (_http_outgoing.js:615:10)
    at Request.write (/path/to/thing/node_modules/request/request.js:1399:27)
    at FormData.ondata (internal/streams/legacy.js:16:26)
    at emitOne (events.js:115:13)
    at FormData.emit (events.js:210:7)
    at FormData.CombinedStream.write (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:118:8)
    at FormData.CombinedStream._pipeNext (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:106:8)
    at FormData.CombinedStream._getNext (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:79:10)
    at FormData.CombinedStream._pipeNext (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:107:8)
    at FormData.CombinedStream._getNext (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:79:10)
    at FormData.CombinedStream._pipeNext (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:107:8)
    at FormData.<anonymous> (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:91:10)
    at FormData.<anonymous> (/path/to/thing/node_modules/form-data/lib/form_data.js:273:5)
    at FormData.CombinedStream._getNext (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:84:3)
    at FormData.CombinedStream._pipeNext (/path/to/thing/node_modules/combined-stream/lib/combined_stream.js:107:8)
```

#### Attachments:

Sample app to reproduce:

```javascript
const { SLACK_TOKEN } = process.env;
const { WebClient } = require('@slack/client');

const web = new WebClient(SLACK_TOKEN);

web.reactions.get({ file: 'STRING', full: true });
```

### PR DESCRIPTION
**Pull request by aoberoi**
###  Summary

based on #513 - as this PR states, this should allow methods like `users.setPhoto` to also work properly.

fixes #452 

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

File: src/WebClient.spec.js [hunk 1]
```diff
@@ -201,23 +215,28 @@ describe('WebClient', function () {
...
+
+            assert.lengthOf(parts.fields, 2);
+            assert.deepInclude(parts.fields, { fieldname: 'token', value: token });
+            assert.deepInclude(parts.fields, { fieldname: 'filename', value: 'train.jpg' });
           });
       });
 
-      // Reactivate this test once we find out if the workaround in the test case before is necessary
-      it('should log a warning when file is a Buffer and there is no filename', function () {
+      // TODO: re-enable this once we decide if we want to generate a random name or if we want to log a warning
```
Comment (aoberoi) [inline src/WebClient.spec.js]: an open point we should resolve before merging this.

**Comment by aoberoi** [comment]
blocker: i just tried using this implementation with a ReadableStream from `fs.createReadStream()` and it failed with the following error:

```
   { ok: false,
     error: 'no_file_data',
     acceptedScopes: [ 'files:write:user', 'post' ] } }
```

**Comment by aoberoi** [comment]
okay, i've learned an interesting thing. the following is true for `files.upload` (i haven't checked against other methods, but i will):

if in the multipart body, the part that contains the binary data does not specify a `filename` attribute in the header, the upload will fail. if another part contains a `filename` that will override whatever was specified in the `filename` attribute of the first part.

**Comment by aoberoi** [comment]
and the same is true for `users.setPhoto`. that is that if no `filename` attribute is specified, it will fail. in this case, there's no need for a `filename` attribute, since the filename is generated by Slack.

**Comment by aoberoi** [comment]
side note: love it when you implement a new feature with more deletions than additions 💯

File: src/WebClient.ts [hunk 2]
```diff
@@ -656,6 +651,8 @@ export interface WebAPIHTTPError extends CodedError {
  * Helpers
  */
 
+const defaultFilename = 'Untitled';
```
Comment (Roach) [inline src/WebClient.ts]: Should we add a note to the docs for this?

Comment (aoberoi) [reply src/WebClient.ts]: heh, if we had generated docs for the individual methods at all, then i'd say yes.

**Pull request merged**